### PR TITLE
[#378] feat: introduce storage manager selector

### PR DESCRIPTION
**Pull request by zuston**
### What changes were proposed in this pull request?

1. introduce storage manager selector to support more selector strategy for `MultiStorageManager`
2. Introduce the conf of `rss.server.multistorage.cold.storage.preferred.factor` to support different flush strategy, like I hope huge partition directly flushed to HDFS and normal partition could be flushed to DISK when single buffer flush is enabled. 

### Why are the changes needed?
Solving the problem mentioned in https://github.com/apache/incubator-uniffle/issues/378#issuecomment-1373447729. 

In current codebase, when encountering huge partition, if single buffer flush is enabled, the normal partition data will be flush to HDFS(I don't hope so, because the local disk is free and the flushing speed is faster than HDFS). But if disable single flush buffer, the huge partition event before marking as huge partition may be big, which cause the slow flushing and then cause requiring allocated buffer failed.

Based on above problems, this PR is to make single event carrying with 100 mb flushed into HDFS or local file leveraging the conf of `rss.server.multistorage.cold.storage.preferred.factor`

### Does this PR introduce _any_ user-facing change?
Yes. Doc will be updated later.

### How was this patch tested?
1. UTs

### CODE REVIEW

**Comment by zuston** [comment]
PTAL @jerqi @xianjingfeng . This is a remaining improvement for #378

File: server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java [hunk 1]
```diff
@@ -226,43 +226,50 @@ void flushSingleBufferIfNecessary(
       int startPartition,
       int endPartition,
       TripleFunction<String, Integer, Integer, Long> getPartitionDataSizeFunc) {
+    boolean isHugePartition =
+        getPartitionDataSizeFunc != null
+            && getPartitionDataSizeFunc.accept(appId, shuffleId, partitionId) > hugePartitionSizeThreshold;
+
     // When we use multi storage and trigger single buffer flush, the buffer size should be bigger
     // than rss.server.flush.cold.storage.threshold.size, otherwise cold storage will be useless.
     if (this.bufferFlushEnabled && buffer.getSize() > this.bufferFlushThreshold) {
```
Comment (xianjingfeng) [inline server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java:235]:
```suggestion
    if ((this.bufferFlushEnabled || isHugePartition) && buffer.getSize() > this.bufferFlushThreshold) {
```
WDYT?

File: server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java [hunk 2]
```diff
@@ -56,6 +57,44 @@ public void selectStorageManagerTest() {
...
+    ShuffleServerConf conf = new ShuffleServerConf();
+    conf.setLong(ShuffleServerConf.FLUSH_COLD_STORAGE_THRESHOLD_SIZE, 10000L);
+    conf.set(ShuffleServerConf.RSS_STORAGE_BASE_PATH, Arrays.asList("test"));
+    conf.setLong(ShuffleServerConf.DISK_CAPACITY, 10000L);
+    conf.setString(ShuffleServerConf.RSS_STORAGE_TYPE, StorageType.MEMORY_LOCALFILE_HDFS.name());
+    conf.setString(ShuffleServerConf.MULTISTORAGE_FALLBACK_STRATEGY_CLASS,
+        RotateStorageManagerFallbackStrategy.class.getCanonicalName());
+    conf.set(
+        ShuffleServerConf.MULTISTORAGE_SELECTOR_COLD_STORAGE_PREFER,
+        StorageManagerSelector.ColdStoragePreferredFactor.HUGE_PARTITION
```
Comment (xianjingfeng) [inline server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java:71]: I am not sure is there any problem if change this to conf.setString('rss.server.multistorage.cold.storage.preferred.factor', "HUGE_PARTITION").

File: server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  public DefaultStorageManagerSelector(
+      StorageManager warmStorageManager,
+      StorageManager coldStorageManager,
+      AbstractStorageManagerFallbackStrategy fallbackStrategy,
+      RssConf rssConf) {
+    this.coldStorageManager = coldStorageManager;
+    this.warmStorageManager = warmStorageManager;
+    this.fallbackStrategy = fallbackStrategy;
+    this.flushColdStorageThresholdSize = rssConf.get(FLUSH_COLD_STORAGE_THRESHOLD_SIZE);
+    this.preferColdStorageFactor = rssConf.get(MULTISTORAGE_SELECTOR_COLD_STORAGE_PREFER);
```
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java:44]: Yes, it’s another optional implementation. Let me think twice.

File: server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java (hunk 2 above)
Comment (zuston) [reply server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java]: Could you describe more about your concern?

File: server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java (hunk 1 above)
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java:235]: Yes. It’s Ok

File: server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java (hunk 2 above)
Comment (xianjingfeng) [reply server/src/test/java/org/apache/uniffle/server/storage/MultiStorageManagerTest.java]: I worry too much, please ignore it.

File: server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java:44]: I still think the default storage manager selector is enough to cover this case. If we have more complex requirements, the pluggable storage manager selector could be supported in the future.

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 4]
```diff
@@ -671,7 +671,7 @@ public ShuffleTaskInfo getShuffleTaskInfo(String appId) {
 
   private void triggerFlush() {
     synchronized (this.shuffleBufferManager) {
-      this.shuffleBufferManager.flushIfNecessary();
+      this.shuffleBufferManager.flushIfNecessary(this::getPartitionDataSize);
```
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:674]: The `shuffleTaskManager` is initialized after creating the instance of `ShuffleBufferManager`, emm, we'd better to pass the `ShuffleServer` into `ShuffleBufferManager`. But it still looks unclear, especially for the dependency of different managers. I have no idea on this.

**Comment by zuston** [comment]
> I'd like to check whether the CI workflow for operator is worked as expected or not.

Done

File: server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java (hunk 1 above)
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java]: Done

File: server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/storage/multi/DefaultStorageManagerSelector.java:44]: Let's refactor in the future.

File: server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java [hunk 5]
```diff
@@ -76,7 +78,12 @@ public class ShuffleBufferManager {
   // appId -> shuffleId -> shuffle size in buffer
   protected Map<String, Map<Integer, AtomicLong>> shuffleSizeMap = Maps.newConcurrentMap();
 
-  public ShuffleBufferManager(ShuffleServerConf conf, ShuffleFlushManager shuffleFlushManager) {
+  public ShuffleBufferManager(ShuffleServerConf conf, ShuffleServer shuffleServer) {
+    this(conf, shuffleServer.getShuffleFlushManager(), shuffleServer.getShuffleTaskManager());
```
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java]: Yes. Fixed

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 6]
```diff
@@ -263,6 +264,15 @@ public class ShuffleServerConf extends RssBaseConf {
       .defaultValue(64L * 1024L * 1024L)
       .withDescription("For multistorage, the event size exceed this value, flush data  to cold storage");
 
+  public static final ConfigOption<StorageManagerSelector.ColdStoragePreferredFactor>
+      MULTISTORAGE_SELECTOR_COLD_STORAGE_PREFER = ConfigOptions
+      .key("rss.server.multistorage.cold.storage.preferred.factor")
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:269]: Could you give this config option a better name? It is more like a strategy in our system.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 7]
```diff
@@ -263,6 +264,15 @@ public class ShuffleServerConf extends RssBaseConf {
...
+      .enumType(StorageManagerSelector.ColdStoragePreferredFactor.class)
+      .defaultValue(StorageManagerSelector.ColdStoragePreferredFactor.HUGE_EVENT)
+      .withDescription("The cold storage preferred factor for multiple storage manager. Only the value is "
+          + StorageManagerSelector.ColdStoragePreferredFactor.HUGE_EVENT + ", the conf of "
```
[the remark this answers is not in the document]
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:273]: Got it.

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java (hunk 6 above)
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:269]: WDYT?

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:269]: rss.server.multistorage.selector.strategy

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:269]: This has been discussed in above conversation. https://github.com/apache/incubator-uniffle/pull/621#discussion_r1111945895. If you prefer, I will refactor it.

Comment (jerqi) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:269]: It's better to refactor ... it's more clear.

**Review by jerqi — APPROVED** [review]
LGTM

**Pull request merged**